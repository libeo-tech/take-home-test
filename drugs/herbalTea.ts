import { Drug, DrugName } from "./drug";

export class HerbalTea extends Drug {
    constructor(expiresIn: number, benefit: number) {
        super(DrugName.HERBAL_TEA, expiresIn, benefit);
    }

    public updateDrug(): void {
        if (this.isExpired) {
            this.incrementBenefit(2);
            return;
        }
        
        this.decreaseExpiresIn();
        this.incrementBenefit();
    }
}