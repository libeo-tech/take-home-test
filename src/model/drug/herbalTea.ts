import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class HerbalTea extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.HERBAL_TEA, expiresIn, benefit);
  }

  public updateValues() {
    if (this.isExpired()) {
      this.benefit += 2;
    } else {
      this.benefit += 1;
    }
    this.expiresIn -= 1;

    return this;
  }
}
