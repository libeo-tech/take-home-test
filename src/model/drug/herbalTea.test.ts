import { HerbalTea } from './herbalTea';

describe('Herbal Tea', () => {
  let herbalTea: HerbalTea;

  beforeEach(() => {
    herbalTea = new HerbalTea(30, 5);
  });

  it('should not allow for a negative benefit', () => {
    herbalTea.benefit = -5;
    expect(herbalTea).toMatchSnapshot();
  });

  it('should not allow for a benefit greater than 50', () => {
    herbalTea.benefit = 51;
    expect(herbalTea).toMatchSnapshot();
  });

  it('should return expiration status', () => {
    expect(herbalTea.isExpired()).toBe(false);
    herbalTea.expiresIn = 0;
    expect(herbalTea.isExpired()).toBe(true);
  });

  it('should increase benefit with time', () => {
    herbalTea.updateValues();
    expect(herbalTea).toMatchSnapshot();
  });

  it('should increase benefit twice as much if expired', () => {
    herbalTea.expiresIn = 0;
    herbalTea.updateValues();
    expect(herbalTea).toMatchSnapshot();
  });

  it('should be serializable to JSON', () => {
    expect(herbalTea.toJSON()).toMatchSnapshot();
  });

  it('should be serializable to string', () => {
    expect(herbalTea.toString()).toMatchSnapshot();
  });
});
