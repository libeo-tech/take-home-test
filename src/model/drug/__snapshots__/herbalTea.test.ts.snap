// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Herbal Tea should be serializable to JSON 1`] = `
Object {
  "benefit": 5,
  "expiresIn": 30,
  "name": "Herbal Tea",
}
`;

exports[`Herbal Tea should be serializable to string 1`] = `"{\\"name\\":\\"Herbal Tea\\",\\"expiresIn\\":30,\\"benefit\\":5}"`;

exports[`Herbal Tea should increase benefit twice as much if expired 1`] = `
Object {
  "benefit": 7,
  "expiresIn": -1,
  "name": "Herbal Tea",
}
`;

exports[`Herbal Tea should increase benefit with time 1`] = `
Object {
  "benefit": 6,
  "expiresIn": 29,
  "name": "Herbal Tea",
}
`;

exports[`Herbal Tea should not allow for a benefit greater than 50 1`] = `
Object {
  "benefit": 50,
  "expiresIn": 30,
  "name": "Herbal Tea",
}
`;

exports[`Herbal Tea should not allow for a negative benefit 1`] = `
Object {
  "benefit": 0,
  "expiresIn": 30,
  "name": "Herbal Tea",
}
`;
